### PR TITLE
New validate_list_of()

### LINKED ISSUE #193: Implement and export validate_list_of()
https://github.com/tidyverse/blob/pull/11#discussion_r255502070

### PR DESCRIPTION
**Pull request by krlmlr**
using the constructor, which needs to compute `ptype` anyway.

Closes #193.

### CODE REVIEW

File: R/list_of.R [hunk 1]
```diff
@@ -65,6 +65,15 @@ new_list_of <- function(x = list(), ptype = logical(), ..., class = character())
   new_vctr(x, ..., ptype = ptype, class = c(class, "vctrs_list_of"))
 }
 
+#' @export
+#' @rdname list_of
+validate_list_of <- function(x) {
+  # Type checking is a side effect of the constructor
+  list_of(!!!x, .ptype = attr(x, "ptype"))
```
Comment (hadley) [inline R/list_of.R]: Why not just inline the code and do `map(x, vec_cast, to = ptype)`?  Also the code currently doesn't check that `x` is a list

**Comment by krlmlr** [comment]
Done. Can we please get `walk()` in our `compat-purrr.R` ?

File: R/compat-purrr.R [hunk 2]
```diff
@@ -30,6 +30,17 @@ map_cpl <- function(.x, .f, ...) {
   map_mold(.x, .f, complex(1), ...)
 }
 
+walk <- function(.x, .f, ...) {
```
Comment (hadley) [inline R/compat-purrr.R]: I think you can simplify to `map(.x, .f, ...); invisible(.x)`

Comment (krlmlr) [reply R/compat-purrr.R]: Done.

**Comment by hadley** [comment]
Thanks!

**Pull request merged**